src/org/ods/component/HelmDeploymentStrategy.groovy

```diff
@@ -0,0 +1,255 @@
+            config.helmEnvBasedValuesFiles = []
+        }
+        if (!config.containsKey('helmDefaultFlags')) {
+            config.helmDefaultFlags = ['--install', '--atomic']
+        }
+        if (!config.containsKey('helmAdditionalFlags')) {
+            config.helmAdditionalFlags = []
+        }
+        if (!config.containsKey('helmDiff')) {
+            config.helmDiff = true
+        }
+        if (!config.helmPrivateKeyCredentialsId) {
+            config.helmPrivateKeyCredentialsId = "${context.cdProject}-helm-private-key"
+        }
+        this.script = script
+        this.context = context
+        this.logger = logger
+        this.steps = new PipelineSteps(script)
+
+        this.options = new RolloutOpenShiftDeploymentOptions(config)
+        this.openShift = openShift
+        this.jenkins = jenkins
+    }
+
+    @Override
+    Map<String, List<PodData>> deploy() {
+        if (!context.environment) {
+            logger.warn 'Skipping because of empty (target) environment ...'
+            return [:]
+        }
+
+        def deploymentResources = openShift.getResourcesForComponent(
+            context.targetProject, DEPLOYMENT_KINDS, options.selector
+        )
+        // Tag images which have been built in this pipeline from cd project into target project
+        retagImages(context.targetProject, getBuiltImages())
+
+        logger.info "Rolling out ${context.componentId} with HELM, selector: ${options.selector}"
+        helmUpgrade(context.targetProject)
+
+        deploymentResources = openShift.getResourcesForComponent(
+            context.targetProject, DEPLOYMENT_KINDS, options.selector
+        )
+        logger.info("${this.class.name} -- DEPLOYMENT RESOURCES")
+        logger.info(
+            JsonOutput.prettyPrint(
+                JsonOutput.toJson(deploymentResources)))
+
+
+        Map<String,String> labels = showMandatoryLabels()
+        applyMandatoryLabels(labels, deploymentResources)
+
+        // // FIXME: pauseRollouts is non trivial to determine!
+        // // we assume that Helm does "Deployment" that should work for most
+        // // cases since they don't have triggers.
+        // metadataSvc.updateMetadata(false, deploymentResources)
+        def rolloutData = rollout(deploymentResources) ?: [:]
+        logger.info(groovy.json.JsonOutput.prettyPrint(groovy.json.JsonOutput.toJson(rolloutData)))
+        return rolloutData
+    }
+
+    private void applyMandatoryLabels(Map<String, String> labels, Map<String, List<String>> deploymentResources) {
+        logger.debug("${this.class.name} -- MANDATORY LABELS")
+        logger.debug(
+            JsonOutput.prettyPrint(
+                JsonOutput.toJson(labels)))
+        logger.debug("${this.class.name} -- MANDATORY LABELS (cleaned)")
+        logger.debug(
+            JsonOutput.prettyPrint(
+                JsonOutput.toJson(labels)))
+        labels.remove { it -> it.value == null }
```

Comment:
Null values are used to mark labels for deletion by OpenShiftResourceMetadata.updateMetadata(). In some cases, one or more optional labels may not apply anymore and need to be deleted. If you just ignore them, we may end up with wrong labels set.
Actually OpenShiftResourceMetadata.updateMetadata should do its job correctly, regardless of using Tailor or Helm. Is there any difference that makes it unsuitable for the Helm strategy? It should be possible to adapt it.

Reply:
Sure, I will schedule a meeting

Reply:
@jafarre-bi were you able to resolve your questions in the meeting?

Reply:
@serverhorror ok, we'll discuss internally!